File: conformance/runner/kotlin/src/main/kotlin/com/basecamp/hey/conformance/Assertions.kt
package com.basecamp.hey.conformance

import com.basecamp.hey.HeyException
import com.basecamp.hey.nextLink
import com.basecamp.hey.generated.Routes
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.longOrNull
import java.net.URI

class AssertionFailure(message: String) : RuntimeException(message)

/** One case after it ran: what the SDK answered and what the mock server saw. */
class Run(
    val case: TestCase,
    val outcome: Result<Outcome>,
    val recorded: Recorded,
    val baseUrl: String,
)

fun checkAll(run: Run) {
    for (assertion in run.case.assertions) check(run, assertion)
}

private fun fail(message: String): Nothing = throw AssertionFailure(message)

private fun check(run: Run, assertion: Assertion) {
    when (assertion.type) {
        "requestCount" -> checkRequestCount(run, assertion)
        "delayBetweenRequests" -> checkDelayBetweenRequests(run, assertion)
        "noError" -> checkNoError(run)
        "errorCode" -> checkErrorCode(run, assertion)
        "errorField" -> checkErrorField(run, assertion)
        "statusCode" -> checkStatusCode(run, assertion)
        "requestPath" -> checkRequestPath(run, assertion, Which.FIRST)
        "lastRequestPath" -> checkRequestPath(run, assertion, Which.LAST)
        "requestMethod" -> checkRequestMethod(run, assertion)
        "requestQuery" -> checkRequestQuery(run, assertion, Which.FIRST)
        "lastRequestQuery" -> checkRequestQuery(run, assertion, Which.LAST)
        "requestBody" -> checkRequestBody(run, assertion)
        "requestForm" -> checkRequestForm(run, assertion, Which.FIRST)
        "lastRequestForm" -> checkRequestForm(run, assertion, Which.LAST)
        "headerPresent" -> checkHeaderPresent(run, assertion)
        "lastRequestHeader" -> checkLastRequestHeader(run, assertion)
        "responseMeta" -> checkResponseMeta(run, assertion)
        "urlOrigin" -> checkUrlOrigin(run, assertion)
        "responseBody" -> checkResponseBody(run, assertion)
        else -> fail("Unknown assertion type: ${assertion.type}")
    }
}

private fun checkRequestCount(run: Run, assertion: Assertion) {
    val expected = expectedInt(assertion, "requestCount")
    if (run.recorded.count.toLong() != expected) fail("Expected $expected requests, got ${run.recorded.count}")
}

private fun checkDelayBetweenRequests(run: Run, assertion: Assertion) {
    if (run.recorded.times.size < 2) return
    val delay = (run.recorded.times[1] - run.recorded.times[0]) / 1_000_000
    val minimum = assertion.min.toLong()
    if (delay < minimum) fail("Expected delay >= ${minimum}ms, got ${delay}ms")
}

private fun emptyStatuses(operation: String): List<Int> = Routes.ALL.firstOrNull { it.id == operation }?.emptyOn.orEmpty()

private fun checkNoError(run: Run) {
    val error = run.outcome.exceptionOrNull() ?: return
    if (lastStatus(run) in emptyStatuses(run.case.operation)) {
        fail("Expected no error, got: $error (the route treats ${lastStatus(run)} as empty for ${run.case.operation})")
    }
    fail("Expected no error, got: $error")
}

private fun checkErrorCode(run: Run, assertion: Assertion) {
    val expected = expectedString(assertion, "errorCode")
    val error = run.outcome.exceptionOrNull() ?: fail("Expected error code \"$expected\", but got no error")
    val code = (error as? HeyException)?.code ?: fail("Expected error code \"$expected\", got a non-SDK failure: $error")
    if (code != expected) fail("Expected error code \"$expected\", got \"$code\"")
}

private fun checkErrorField(run: Run, assertion: Assertion) {
    val error = run.outcome.exceptionOrNull() as? HeyException
        ?: fail("Expected error field \"${assertion.path}\", but got no error")
    when (assertion.path) {
        "httpStatus" -> {
            val expected = expectedInt(assertion, "errorField.httpStatus")
            val actual = error.httpStatus ?: 0
            if (actual.toLong() != expected) fail("Expected error httpStatus $expected, got $actual")
        }
        "retryable" -> {
            val expected = expectedBool(assertion, "errorField.retryable")
            if (error.retryable != expected) fail("Expected error retryable=$expected, got ${error.retryable}")
        }
        "requestId" -> {
            val expected = expectedString(assertion, "errorField.requestId")
            val actual = error.requestId ?: ""
            if (actual != expected) fail("Expected error requestId \"$expected\", got \"$actual\"")
        }
        else -> fail("Unknown error field: ${assertion.path}")
    }
}

/** A failure has to carry the status on the error itself; only a success reads it off the server. */
private fun checkStatusCode(run: Run, assertion: Assertion) {
    val expected = expectedInt(assertion, "statusCode")
    val error = run.outcome.exceptionOrNull()
    val actual = if (error != null) {
        (error as? HeyException)?.httpStatus?.takeIf { it > 0 }
            ?: fail("Expected status code $expected, but the SDK error carries no HTTP status: $error")
    } else {
        lastStatus(run)
    }
    if (actual.toLong() != expected) fail("Expected status code $expected, got $actual")
}

private fun lastStatus(run: Run): Int = run.recorded.statuses.lastOrNull() ?: 0

private fun checkRequestPath(run: Run, assertion: Assertion, which: Which) {
    val raw = expectedString(assertion, "requestPath")
    val expected = if (run.case.asksForHtml) raw else withJsonExtension(raw)
    val actual = which.pick(run.recorded.paths) ?: fail(noRequests())
    if (actual != expected) fail("Expected ${which.label} request path \"$expected\", got \"$actual\"")
}

/**
 * HEY answers JSON to paths ending in `.json`, and this SDK puts the extension back on a
 * path whose last segment has none, so the expected path is normalised the same way. A page
 * HEY serves as HTML is the exception: the SDK asks for it as written.
 */
fun withJsonExtension(path: String): String {
    val lastSegment = path.substringAfterLast('/')
    return if (path.isEmpty() || path.endsWith("/") || lastSegment.contains('.')) path else "$path.json"
}

private fun checkRequestMethod(run: Run, assertion: Assertion) {
    val expected = expectedString(assertion, "requestMethod")
    val actual = run.recorded.methods.firstOrNull() ?: fail(noRequests())
    if (actual != expected) fail("Expected request method \"$expected\", got \"$actual\"")
}

private fun checkRequestQuery(run: Run, assertion: Assertion, which: Which) {
    val expected = expectedObject(assertion, assertion.type)
    val query = which.pick(run.recorded.queries) ?: fail(noRequests())
    for ((name, want) in expected) {
        val got = query.firstOrNull { it.first == name }?.second
        when {
            want is JsonNull && got == null -> {}
            want is JsonNull -> fail("Expected ${which.label} query param \"$name\" to be absent, got \"$got\"")
            got == null -> fail("Expected ${which.label} query param $name=${display(want)}, got \"\"")
            got != display(want) -> fail("Expected ${which.label} query param $name=${display(want)}, got \"$got\"")
        }
    }
}

private fun checkRequestBody(run: Run, assertion: Assertion) {
    val expected = expectedObject(assertion, "requestBody")
    val raw = run.recorded.bodies.firstOrNull() ?: fail(noRequests())
    val body: JsonElement = if (raw.isEmpty()) {
        JsonNull
    } else {
        runCatching { kotlinx.serialization.json.Json.parseToJsonElement(raw.decodeToString()) }
            .getOrElse { fail("requestBody: request body is not JSON: $it") }
    }
    for ((path, want) in expected) {
        val got = lookup(body, path)
        when {
            want is JsonNull && got == null -> {}
            want is JsonNull -> fail("Expected body key \"$path\" to be absent, got ${display(got!!)}")
            got == null -> fail("Expected body key \"$path\" = ${display(want)}, but it is absent")
            !valuesMatch(want, got) -> fail("Expected body key \"$path\" = ${display(want)}, got ${display(got)}")
        }
    }
}

private fun checkRequestForm(run: Run, assertion: Assertion, which: Which) {
    val expected = expectedObject(assertion, assertion.type)
    val raw = which.pick(run.recorded.bodies) ?: fail(noRequests())
    val fields = queryPairs(raw.decodeToString())
    for ((name, want) in expected) {
        val got = fields.firstOrNull { it.first == name }?.second
        when {
            want is JsonNull && got == null -> {}
            want is JsonNull -> fail("${assertion.type}: expected form field \"$name\" to be absent, got \"$got\"")
            got == null -> fail("${assertion.type}: expected form field \"$name\" = ${display(want)}, but it is absent")
            got != display(want) -> fail("${assertion.type}: expected form field \"$name\" = ${display(want)}, got \"$got\"")
        }
    }
}

private fun checkHeaderPresent(run: Run, assertion: Assertion) {
    val name = assertion.path
    if (run.recorded.headers.isEmpty()) fail("Expected request with header \"$name\", but no requests were recorded")
    if (run.recorded.header(0, name).isNullOrEmpty()) fail("Expected header \"$name\" to be present, but it was not")
}

private fun checkLastRequestHeader(run: Run, assertion: Assertion) {
    val name = assertion.path
    val expected = expectedString(assertion, "lastRequestHeader")
    if (run.recorded.headers.isEmpty()) fail("Expected request with header \"$name\", but no requests were recorded")
    val actual = run.recorded.header(run.recorded.headers.size - 1, name) ?: ""
    if (actual != expected) fail("Expected last request header \"$name\" = \"$expected\", got \"$actual\"")
}

private fun checkResponseMeta(run: Run, assertion: Assertion) {
    val page = run.outcome.getOrNull() as? Outcome.Page
        ?: fail("Expected a paginated result to read responseMeta.${assertion.path} from")
    when (assertion.path) {
        "totalCount" -> {
            val expected = expectedInt(assertion, "responseMeta.totalCount")
            val actual = page.totalCount ?: fail("X-Total-Count header not present in response")
            if (actual != expected) fail("Expected X-Total-Count=$expected, got $actual")
        }
        "nextPage" -> {
            val expected = expectedString(assertion, "responseMeta.nextPage")
            val actual = page.nextPage ?: fail("Link header does not contain a valid next URL")
            if (actual != expected) fail("Expected next page \"$expected\", got \"$actual\"")
        }
        else -> fail("Unknown responseMeta path: ${assertion.path}")
    }
}

private fun checkUrlOrigin(run: Run, assertion: Assertion) {
    val expected = expectedString(assertion, "urlOrigin")
    if (expected != "rejected") fail("urlOrigin: unsupported expected value \"$expected\" (only \"rejected\" is supported)")
    val link = run.recorded.links.lastOrNull() ?: fail("No Link header in response to validate origin")
    if (link.isNullOrEmpty()) fail("No Link header in response to validate origin")
    val target = nextLink(link) ?: fail("No next URL found in Link header: $link")
    val server = URI(run.baseUrl)
    val next = runCatching { URI(target) }.getOrNull()?.takeIf { it.isAbsolute }
        ?: fail("Expected cross-origin Link URL for rejection test, but got relative URL: $target")
    if (sameOrigin(next, server)) fail("Expected cross-origin Link URL for rejection test, but $target has same origin as server")
    checkNextPageRefused(run)
}

private fun port(uri: URI): Int = if (uri.port >= 0) uri.port else if (uri.scheme.equals("https", true)) 443 else 80

private fun sameOrigin(a: URI, b: URI): Boolean =
    a.scheme.equals(b.scheme, ignoreCase = true) && a.host.orEmpty().equals(b.host.orEmpty(), ignoreCase = true) && port(a) == port(b)

private fun checkNextPageRefused(run: Run) {
    val page = run.outcome.getOrElse { fail("Expected a paginated result to read the next page from, got: $it") } as? Outcome.Page
        ?: fail("Expected a paginated result to read the next page from")
    val check = page.nextUrlCheck ?: fail("Expected a paginated result to read the next page from")
    val error = check.exceptionOrNull() ?: fail("Expected the cross-origin next page to be refused, but the SDK followed it")
    val code = (error as? HeyException)?.code
    if (code != HeyException.CODE_USAGE) fail("Expected the cross-origin next page to be refused as a usage error, got ${code ?: error}")
}

private fun checkResponseBody(run: Run, assertion: Assertion) {
    val path = assertion.path
    val outcome = run.outcome.getOrElse { fail("Expected responseBody.$path, got: $it") }
    val body = outcome.body() ?: fail("Expected responseBody.$path, but no response body captured")
    val actual = lookup(body, path) ?: fail("Expected responseBody.$path, but field not present")
    if (!valuesMatch(assertion.expected, actual)) fail("Expected responseBody.$path = ${display(assertion.expected)}, got ${display(actual)}")
}

private enum class Which(val label: String) {
    FIRST("first"),
    LAST("last"),
    ;

    fun <T> pick(values: List<T>): T? = if (this == FIRST) values.firstOrNull() else values.lastOrNull()
}

private fun noRequests(): String = "Expected a request, but none were recorded"

private fun expectedInt(assertion: Assertion, label: String): Long =
    (assertion.expected as? JsonPrimitive)?.longOrNull ?: fail("$label: expected an integer, got ${display(assertion.expected)}")

private fun expectedBool(assertion: Assertion, label: String): Boolean =
    (assertion.expected as? JsonPrimitive)?.booleanOrNull ?: fail("$label: expected a bool, got ${display(assertion.expected)}")

private fun expectedString(assertion: Assertion, label: String): String =
    (assertion.expected as? JsonPrimitive)?.takeIf { it.isString }?.content ?: fail("$label: expected a string, got ${display(assertion.expected)}")

private fun expectedObject(assertion: Assertion, label: String): JsonObject =
    assertion.expected as? JsonObject ?: fail("$label: expected an object, got ${display(assertion.expected)}")

/** Walks a JSON value by a dot-separated path, reading integer segments as array indexes. */
fun lookup(value: JsonElement, path: String): JsonElement? {
    var current: JsonElement = value
    for (segment in path.split('.')) {
        current = when (current) {
            is JsonObject -> current[segment] ?: return null
            is JsonArray -> segment.toIntOrNull()?.let { current.getOrNull(it) } ?: return null
            else -> return null
        }
    }
    return current
}

fun valuesMatch(expected: JsonElement, actual: JsonElement): Boolean {
    val expectedLong = (expected as? JsonPrimitive)?.longOrNull
    val actualLong = (actual as? JsonPrimitive)?.longOrNull
    if (expectedLong != null && actualLong != null) return expectedLong == actualLong
    val expectedDouble = (expected as? JsonPrimitive)?.doubleOrNull
    val actualDouble = (actual as? JsonPrimitive)?.doubleOrNull
    if (expectedDouble != null && actualDouble != null) return expectedDouble == actualDouble
    val expectedBool = (expected as? JsonPrimitive)?.booleanOrNull
    val actualBool = (actual as? JsonPrimitive)?.booleanOrNull
    if (expectedBool != null && actualBool != null) return expectedBool == actualBool
    return display(expected) == display(actual)
}

fun display(value: JsonElement): String = if (value is JsonPrimitive && value.isString) value.content else value.toString()

File: conformance/runner/swift/Sources/ConformanceSupport/Assertions.swift
import Foundation
import Hey

public struct AssertionFailure: Error, CustomStringConvertible {
    public let description: String
    public init(_ description: String) { self.description = description }
}

/// What an operation answered, in a shape the assertions can read: the parsed model as JSON, plus
/// a page's metadata.
public enum Outcome: Sendable {
    case unit
    case json(FixtureJSON)
    /// `nextURLCheck` is what following the next page did, when the case asked.
    case page(value: FixtureJSON, nextPage: String?, totalCount: Int?, nextURLCheck: Result<Void, any Error>?)

    public var body: FixtureJSON? {
        switch self {
        case .unit: return nil
        case let .json(value): return value
        case let .page(value, _, _, _): return value
        }
    }
}

/// One case after it ran: what the SDK answered and what the mock server saw.
public struct Run: Sendable {
    public var testCase: TestCase
    public var outcome: Result<Outcome, any Error>
    public var recorded: Recorded
    public var baseURL: String

    public init(testCase: TestCase, outcome: Result<Outcome, any Error>, recorded: Recorded, baseURL: String) {
        self.testCase = testCase
        self.outcome = outcome
        self.recorded = recorded
        self.baseURL = baseURL
    }

    var error: (any Error)? {
        if case let .failure(error) = outcome { return error }
        return nil
    }
}

public func checkAll(_ run: Run) throws {
    for assertion in run.testCase.assertions {
        try check(run, assertion)
    }
}

private func fail(_ message: String) -> AssertionFailure { AssertionFailure(message) }

private enum Which {
    case first, last

    var label: String { self == .first ? "first" : "last" }

    func pick<T>(_ values: [T]) -> T? { self == .first ? values.first : values.last }
}

private func check(_ run: Run, _ assertion: Assertion) throws {
    switch assertion.type {
    case "requestCount": try checkRequestCount(run, assertion)
    case "delayBetweenRequests": try checkDelayBetweenRequests(run, assertion)
    case "noError": try checkNoError(run)
    case "errorCode": try checkErrorCode(run, assertion)
    case "errorField": try checkErrorField(run, assertion)
    case "statusCode": try checkStatusCode(run, assertion)
    case "requestPath": try checkRequestPath(run, assertion, .first)
    case "lastRequestPath": try checkRequestPath(run, assertion, .last)
    case "requestMethod": try checkRequestMethod(run, assertion)
    case "requestQuery": try checkRequestQuery(run, assertion, .first)
    case "lastRequestQuery": try checkRequestQuery(run, assertion, .last)
    case "requestBody": try checkRequestBody(run, assertion)
    case "requestForm": try checkRequestForm(run, assertion, .first)
    case "lastRequestForm": try checkRequestForm(run, assertion, .last)
    case "headerPresent": try checkHeaderPresent(run, assertion)
    case "lastRequestHeader": try checkLastRequestHeader(run, assertion)
    case "responseMeta": try checkResponseMeta(run, assertion)
    case "urlOrigin": try checkURLOrigin(run, assertion)
    case "responseBody": try checkResponseBody(run, assertion)
    default: throw fail("Unknown assertion type: \(assertion.type)")
    }
}

private func checkRequestCount(_ run: Run, _ assertion: Assertion) throws {
    let expected = try expectedInt(assertion, "requestCount")
    if run.recorded.count != expected { throw fail("Expected \(expected) requests, got \(run.recorded.count)") }
}

/// Every wait between one request and the next, not only the first: a backoff that collapses after
/// the second send is still a broken backoff.
private func checkDelayBetweenRequests(_ run: Run, _ assertion: Assertion) throws {
    let minimum = Int64(assertion.min)
    let times = run.recorded.times
    for index in times.indices.dropFirst() {
        let delay = (Int64(times[index]) - Int64(times[index - 1])) / 1_000_000
        if delay < minimum { throw fail("Expected delay >= \(minimum)ms before request \(index + 1), got \(delay)ms") }
    }
}

private func lastStatus(_ run: Run) -> Int { run.recorded.statuses.last ?? 0 }

private func checkNoError(_ run: Run) throws {
    guard let error = run.error else { return }
    let emptyOn = Routes.all.first { $0.id == run.testCase.operation }?.emptyOn ?? []
    if emptyOn.contains(lastStatus(run)) {
        throw fail("Expected no error, got: \(error) (the route treats \(lastStatus(run)) as empty for \(run.testCase.operation))")
    }
    throw fail("Expected no error, got: \(error)")
}

private func checkErrorCode(_ run: Run, _ assertion: Assertion) throws {
    let expected = try expectedString(assertion, "errorCode")
    guard let error = run.error else { throw fail("Expected error code \"\(expected)\", but got no error") }
    guard let heyError = error as? HeyError else {
        throw fail("Expected error code \"\(expected)\", got a non-SDK failure: \(error)")
    }
    if heyError.code != expected { throw fail("Expected error code \"\(expected)\", got \"\(heyError.code)\"") }
}

private func checkErrorField(_ run: Run, _ assertion: Assertion) throws {
    guard let error = run.error as? HeyError else {
        throw fail("Expected error field \"\(assertion.path)\", but got no error")
    }
    switch assertion.path {
    case "httpStatus":
        let expected = try expectedInt(assertion, "errorField.httpStatus")
        let actual = error.httpStatus ?? 0
        if actual != expected { throw fail("Expected error httpStatus \(expected), got \(actual)") }
    case "retryable":
        let expected = try expectedBool(assertion, "errorField.retryable")
        if error.isRetryable != expected { throw fail("Expected error retryable=\(expected), got \(error.isRetryable)") }
    case "requestId":
        let expected = try expectedString(assertion, "errorField.requestId")
        let actual = error.requestId ?? ""
        if actual != expected { throw fail("Expected error requestId \"\(expected)\", got \"\(actual)\"") }
    default:
        throw fail("Unknown error field: \(assertion.path)")
    }
}

/// A failure has to carry the status on the error itself; only a success reads it off the server.
private func checkStatusCode(_ run: Run, _ assertion: Assertion) throws {
    let expected = try expectedInt(assertion, "statusCode")
    let actual: Int
    if let error = run.error {
        guard let status = (error as? HeyError)?.httpStatus, status > 0 else {
            throw fail("Expected status code \(expected), but the SDK error carries no HTTP status: \(error)")
        }
        actual = status
    } else {
        actual = lastStatus(run)
    }
    if actual != expected { throw fail("Expected status code \(expected), got \(actual)") }
}

private func checkRequestPath(_ run: Run, _ assertion: Assertion, _ which: Which) throws {
    let raw = try expectedString(assertion, "requestPath")
    let expected = run.testCase.asksForHTML ? raw : withJSONExtension(raw)
    guard let actual = which.pick(run.recorded.paths) else { throw fail(noRequests) }
    if actual != expected { throw fail("Expected \(which.label) request path \"\(expected)\", got \"\(actual)\"") }
}

/// HEY answers JSON to paths ending in `.json`, and this SDK puts the extension back on a path
/// whose last segment has none, so the expected path is normalised the same way. A page HEY serves
/// as HTML is the exception: the SDK asks for it as written.
public func withJSONExtension(_ path: String) -> String {
    let lastSegment = path.split(separator: "/", omittingEmptySubsequences: false).last ?? ""
    return path.isEmpty || path.hasSuffix("/") || lastSegment.contains(".") ? path : path + ".json"
}

private func checkRequestMethod(_ run: Run, _ assertion: Assertion) throws {
    let expected = try expectedString(assertion, "requestMethod")
    guard let actual = run.recorded.methods.first else { throw fail(noRequests) }
    if actual != expected { throw fail("Expected request method \"\(expected)\", got \"\(actual)\"") }
}

private func checkRequestQuery(_ run: Run, _ assertion: Assertion, _ which: Which) throws {
    let expected = try expectedObject(assertion, assertion.type)
    guard let query = which.pick(run.recorded.queries) else { throw fail(noRequests) }
    for (name, want) in expected {
        // A scalar the fixture expects once has to be there exactly once: a second copy of an
        // account filter is a request scoped to two accounts, whichever HEY reads.
        let values = query.filter { $0.0 == name }.map(\.1)
        if want == .null {
            if !values.isEmpty {
                throw fail("Expected \(which.label) query param \"\(name)\" to be absent, got \"\(values.joined(separator: "&"))\"")
            }
        } else if values.isEmpty {
            throw fail("Expected \(which.label) query param \(name)=\(display(want)), got \"\"")
        } else if values.count > 1 {
            throw fail("Expected \(which.label) query param \(name)=\(display(want)) once, got it \(values.count) times")
        } else if values[0] != display(want) {
            throw fail("Expected \(which.label) query param \(name)=\(display(want)), got \"\(values[0])\"")
        }
    }
}

private func checkRequestBody(_ run: Run, _ assertion: Assertion) throws {
    let expected = try expectedObject(assertion, "requestBody")
    guard let raw = run.recorded.bodies.first else { throw fail(noRequests) }
    let body: FixtureJSON
    if raw.isEmpty {
        body = .null
    } else {
        do {
            body = try FixtureJSON.parse(String(decoding: raw, as: UTF8.self))
        } catch {
            throw fail("requestBody: request body is not JSON: \(error)")
        }
    }
    for (path, want) in expected {
        let got = lookup(body, path)
        switch (want, got) {
        case (.null, nil): continue
        case let (.null, got?): throw fail("Expected body key \"\(path)\" to be absent, got \(display(got))")
        case (_, nil): throw fail("Expected body key \"\(path)\" = \(display(want)), but it is absent")
        case let (_, got?):
            if !valuesMatch(want, got) { throw fail("Expected body key \"\(path)\" = \(display(want)), got \(display(got))") }
        }
    }
}

private func checkRequestForm(_ run: Run, _ assertion: Assertion, _ which: Which) throws {
    let expected = try expectedObject(assertion, assertion.type)
    guard let raw = which.pick(run.recorded.bodies) else { throw fail(noRequests) }
    let fields = queryPairs(String(decoding: raw, as: UTF8.self))
    for (name, want) in expected {
        // A scalar the fixture expects once has to be there exactly once: a second copy with
        // another value is a form the server may read either way.
        let values = fields.filter { $0.0 == name }.map(\.1)
        let type = assertion.type
        if want == .null {
            if !values.isEmpty {
                throw fail("\(type): expected form field \"\(name)\" to be absent, got \"\(values.joined(separator: "&"))\"")
            }
        } else if values.isEmpty {
            throw fail("\(type): expected form field \"\(name)\" = \(display(want)), but it is absent")
        } else if values.count > 1 {
            throw fail("\(type): expected form field \"\(name)\" = \(display(want)) once, got it \(values.count) times")
        } else if values[0] != display(want) {
            throw fail("\(type): expected form field \"\(name)\" = \(display(want)), got \"\(values[0])\"")
        }
    }
}

private func checkHeaderPresent(_ run: Run, _ assertion: Assertion) throws {
    let name = assertion.path
    if run.recorded.headers.isEmpty { throw fail("Expected request with header \"\(name)\", but no requests were recorded") }
    if (run.recorded.header(0, name) ?? "").isEmpty { throw fail("Expected header \"\(name)\" to be present, but it was not") }
}

private func checkLastRequestHeader(_ run: Run, _ assertion: Assertion) throws {
    let name = assertion.path
    let expected = try expectedString(assertion, "lastRequestHeader")
    if run.recorded.headers.isEmpty { throw fail("Expected request with header \"\(name)\", but no requests were recorded") }
    let actual = run.recorded.header(run.recorded.headers.count - 1, name) ?? ""
    if actual != expected { throw fail("Expected last request header \"\(name)\" = \"\(expected)\", got \"\(actual)\"") }
}

private func checkResponseMeta(_ run: Run, _ assertion: Assertion) throws {
    guard case let .success(.page(_, nextPage, totalCount, _)) = run.outcome else {
        throw fail("Expected a paginated result to read responseMeta.\(assertion.path) from")
    }
    switch assertion.path {
    case "totalCount":
        let expected = try expectedInt(assertion, "responseMeta.totalCount")
        guard let actual = totalCount else { throw fail("X-Total-Count header not present in response") }
        if actual != expected { throw fail("Expected X-Total-Count=\(expected), got \(actual)") }
    case "nextPage":
        let expected = try expectedString(assertion, "responseMeta.nextPage")
        guard let actual = nextPage else { throw fail("Link header does not contain a valid next URL") }
        if actual != expected { throw fail("Expected next page \"\(expected)\", got \"\(actual)\"") }
    default:
        throw fail("Unknown responseMeta path: \(assertion.path)")
    }
}

private func checkURLOrigin(_ run: Run, _ assertion: Assertion) throws {
    let expected = try expectedString(assertion, "urlOrigin")
    if expected != "rejected" {
        throw fail("urlOrigin: unsupported expected value \"\(expected)\" (only \"rejected\" is supported)")
    }
    guard let last = run.recorded.links.last, let link = last, !link.isEmpty else {
        throw fail("No Link header in response to validate origin")
    }
    guard let target = nextLink(link) else { throw fail("No next URL found in Link header: \(link)") }
    guard let next = URLComponents(string: target), next.scheme != nil, next.host != nil else {
        throw fail("Expected cross-origin Link URL for rejection test, but got relative URL: \(target)")
    }
    if let server = URLComponents(string: run.baseURL), sameOrigin(next, server) {
        throw fail("Expected cross-origin Link URL for rejection test, but \(target) has same origin as server")
    }
    try checkNextPageRefused(run)
}

private func port(_ url: URLComponents) -> Int {
    url.port ?? (url.scheme?.lowercased() == "https" ? 443 : 80)
}

private func sameOrigin(_ a: URLComponents, _ b: URLComponents) -> Bool {
    a.scheme?.lowercased() == b.scheme?.lowercased() && (a.host ?? "").lowercased() == (b.host ?? "").lowercased()
        && port(a) == port(b)
}

private func checkNextPageRefused(_ run: Run) throws {
    let outcome: Outcome
    switch run.outcome {
    case let .failure(error): throw fail("Expected a paginated result to read the next page from, got: \(error)")
    case let .success(value): outcome = value
    }
    guard case let .page(_, _, _, check?) = outcome else { throw fail("Expected a paginated result to read the next page from") }
    guard case let .failure(error) = check else {
        throw fail("Expected the cross-origin next page to be refused, but the SDK followed it")
    }
    let code = (error as? HeyError)?.code
    if code != HeyError.codeUsage {
        throw fail("Expected the cross-origin next page to be refused as a usage error, got \(code ?? "\(error)")")
    }
}

private func checkResponseBody(_ run: Run, _ assertion: Assertion) throws {
    let path = assertion.path
    let outcome: Outcome
    switch run.outcome {
    case let .failure(error): throw fail("Expected responseBody.\(path), got: \(error)")
    case let .success(value): outcome = value
    }
    guard let body = outcome.body else { throw fail("Expected responseBody.\(path), but no response body captured") }
    guard let actual = lookup(body, path) else { throw fail("Expected responseBody.\(path), but field not present") }
    if !valuesMatch(assertion.expected, actual) {
        throw fail("Expected responseBody.\(path) = \(display(assertion.expected)), got \(display(actual))")
    }
}

private let noRequests = "Expected a request, but none were recorded"

private func expectedInt(_ assertion: Assertion, _ label: String) throws -> Int {
    guard let value = assertion.expected.int else { throw fail("\(label): expected an integer, got \(display(assertion.expected))") }
    return value
}

private func expectedBool(_ assertion: Assertion, _ label: String) throws -> Bool {
    guard let value = assertion.expected.bool else { throw fail("\(label): expected a bool, got \(display(assertion.expected))") }
    return value
}

private func expectedString(_ assertion: Assertion, _ label: String) throws -> String {
    guard let value = assertion.expected.string else { throw fail("\(label): expected a string, got \(display(assertion.expected))") }
    return value
}

private func expectedObject(_ assertion: Assertion, _ label: String) throws -> [(String, FixtureJSON)] {
    guard let value = assertion.expected.object else { throw fail("\(label): expected an object, got \(display(assertion.expected))") }
    return value
}

/// Walks a JSON value by a dot-separated path, reading integer segments as array indexes.
public func lookup(_ value: FixtureJSON, _ path: String) -> FixtureJSON? {
    var current = value
    for segment in path.split(separator: ".", omittingEmptySubsequences: false) {
        switch current {
        case .object:
            guard let next = current[String(segment)] else { return nil }
            current = next
        case let .array(elements):
            guard let index = Int(segment), elements.indices.contains(index) else { return nil }
            current = elements[index]
        default:
            return nil
        }
    }
    return current
}

/// Whether two JSON values are the same value of the same kind. A number and the string of that
/// number are not: an SDK that writes `"1"` where the fixture expects `1` has changed the type on
/// the wire, which is what these fixtures exist to catch. Numbers compare as numbers, so `1` and
/// `1.0` agree.
public func valuesMatch(_ expected: FixtureJSON, _ actual: FixtureJSON) -> Bool {
    switch (expected, actual) {
    case let (.number(a), .number(b)):
        // Exact decimal arithmetic: 1 and 1.0 are the same number, 9007199254740993 and
        // 9007199254740992.0 are not, whatever a double would make of them.
        guard let x = Decimal(string: a, locale: Locale(identifier: "en_US_POSIX")),
              let y = Decimal(string: b, locale: Locale(identifier: "en_US_POSIX"))
        else { return a == b }
        return x == y
    case let (.array(a), .array(b)):
        return a.count == b.count && zip(a, b).allSatisfy { valuesMatch($0, $1) }
    case let (.object(a), .object(b)):
        return a.count == b.count && a.allSatisfy { key, value in
            guard let other = actual[key] else { return false }
            return valuesMatch(value, other)
        }
    default:
        return expected == actual
    }
}

public func display(_ value: FixtureJSON) -> String {
    if case let .string(text) = value { return text }
    return value.text
}
